File: coord64/go/kd_test.go
package kd

import (
	"fmt"
	"math/rand"
	"testing"
)

func TestKDTreeBasic(t *testing.T) {
	tree := Create()
	box1 := Box{0, 0, 10, 10}
	box2 := Box{20, 20, 30, 30}
	box3 := Box{5, 5, 15, 15}

	tree.Insert("item1", box1)
	tree.Insert("item2", box2)
	tree.Insert("item3", box3)

	tree.Delete("item1", box1)
}

func TestKDTreeHardDelete(t *testing.T) {
	tree := Create()
	box1 := Box{0, 0, 10, 10}
	box2 := Box{20, 20, 30, 30}
	box3 := Box{5, 5, 15, 15}

	tree.Insert("item1", box1)
	tree.Insert("item2", box2)
	tree.Insert("item3", box3)

	tree.HardDelete("item1", box1)
}

func TestThousandBoxes(t *testing.T) {
	tree := Create()
	rng := rand.New(rand.NewSource(42))

	var boxes []Box
	for i := 0; i < 1000; i++ {
		x1 := rng.Int63n(10000000000)
		y1 := rng.Int63n(10000000000)
		x2 := x1 + rng.Int63n(100) + 1
		y2 := y1 + rng.Int63n(100) + 1
		b := Box{x1, y1, x2, y2}
		boxes = append(boxes, b)
		tree.Insert(fmt.Sprintf("box%d", i), b)
	}

	if count := tree.Count(); count != 1000 {
		t.Fatalf("Expected 1000 boxes, got %d", count)
	}

	searchArea := Box{0, 0, 5000000000, 5000000000}
	gen := tree.Start(searchArea)
	foundCount := 0
	for {
		_, _, ok := gen.Next()
		if !ok {
			break
		}
		foundCount++
	}
	t.Logf("Found %d boxes in the large search area", foundCount)
}

func TestMillionBoxes(t *testing.T) {
	tree := Create()
	rng := rand.New(rand.NewSource(42))

	var boxesToDelete []Box
	for i := 0; i < 1000000; i++ {
		x1 := rng.Int63n(10000000000)
		y1 := rng.Int63n(10000000000)
		x2 := x1 + rng.Int63n(100) + 1
		y2 := y1 + rng.Int63n(100) + 1
		b := Box{x1, y1, x2, y2}
		
		if i < 1000 {
			boxesToDelete = append(boxesToDelete, b) // Store first 1,000 for hard deletion
		}
		tree.Insert(fmt.Sprintf("box%d", i), b)
	}

	if count := tree.Count(); count != 1000000 {
		t.Fatalf("Expected 1000000 boxes, got %d", count)
	}

	searchArea := Box{0, 0, 5000000000, 5000000000}
	gen := tree.Start(searchArea)
	foundCount := 0
	for {
		_, _, ok := gen.Next()
		if !ok {
			break
		}
		foundCount++
	}
	t.Logf("Found %d boxes in the 0-5,000,000,000 search area", foundCount)

	for i := 0; i < 1000; i++ {
		deleted := tree.HardDelete(fmt.Sprintf("box%d", i), boxesToDelete[i])
		if !deleted {
			t.Errorf("Failed to hard delete box%d", i)
		}
	}

	if count := tree.Count(); count != 999000 {
		t.Fatalf("Expected 999000 boxes after hard deletes, got %d", count)
	}
}

File: coord64/julia/kd_test.jl
using Test
include("kd.jl")
using .KD
using Random

@testset "KDTree Basic & Hard Delete" begin
    tree = Tree{String}()
    box1 = (0, 0, 10, 10)
    box2 = (20, 20, 30, 30)
    box3 = (5, 5, 15, 15)

    KD.insert!(tree, "item1", box1)
    KD.insert!(tree, "item2", box2)
    KD.insert!(tree, "item3", box3)

    @test count_items(tree) == 3
    @test is_member(tree, "item2", box2)

    found = search(tree, (0, 0, 15, 15))
    @test length(found) == 2
    items = [f[1] for f in found]
    @test "item1" in items
    @test "item3" in items

    @test hard_delete!(tree, "item1", box1) == true
    @test count_items(tree) == 2
    @test !is_member(tree, "item1", box1)
    @test is_member(tree, "item3", box3)
end

@testset "Million Boxes Stress Test" begin
    tree = Tree{String}()
    Random.seed!(42)
    
    boxes_to_delete = Tuple{String, Box}[]
    
    # 1. Insert 1,000,000 Boxes
    for i in 1:1_000_000
        x1 = rand(0:99999)
        y1 = rand(0:99999)
        x2 = x1 + rand(1:100)
        y2 = y1 + rand(1:100)
        b = (x1, y1, x2, y2)
        
        name = "box$i"
        if i <= 1000
            push!(boxes_to_delete, (name, b))
        end
        KD.insert!(tree, name, b)
    end
    
    @test count_items(tree) == 1_000_000
    
    # 2. Search a Quarter Sector
    found = search(tree, (0, 0, 50000, 50000))
    println("Found $(length(found)) boxes in the 0-50000 search area")
    @test length(found) > 10000 # Sanity check for data population
    
    # 3. Hard Delete 1,000 Items Deeply Nested in Structure
    for (name, b) in boxes_to_delete
        success = hard_delete!(tree, name, b)
        @test success
    end
    
    @test count_items(tree) == 999_000
end

@testset "Million Boxes 64-bit Stress Test" begin
    tree = Tree{String}()
    Random.seed!(42)
    
    boxes_to_delete = Tuple{String, Box}[]
    
    for i in 1:1_000_000
        x1 = rand(0:10_000_000_000)
        y1 = rand(0:10_000_000_000)
        x2 = x1 + rand(1:100)
        y2 = y1 + rand(1:100)
        b = (x1, y1, x2, y2)
        
        name = "box$i"
        if i <= 1000
            push!(boxes_to_delete, (name, b))
        end
        KD.insert!(tree, name, b)
    end
    
    @test count_items(tree) == 1_000_000
    
    found = search(tree, (0, 0, 5_000_000_000, 5_000_000_000))
    println("Found $(length(found)) boxes in the 0-5,000,000,000 search area")
    @test length(found) > 10000
    
    for (name, b) in boxes_to_delete
        success = hard_delete!(tree, name, b)
        @test success
    end
    
    @test count_items(tree) == 999_000
end
